### PR TITLE
Flow Control

### PR DESCRIPTION
**Pull request by Elad-Laufer**
Please address the following in the **server** implementation:

1. Consider replacing the current DAG validation method with the one presented in the SDK
2. The server _**must**_ support ImportFileJobSpecification as well as the ImportFileRequest as the configuration of the Flow component.

### CODE REVIEW

File: media_platform/service/flow_control_service/flow.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from media_platform.lang.serialization import Serializable, Deserializable
+from media_platform.service.flow_control_service.component import Component
+
+
+class Flow(Serializable, Deserializable):
+    def __init__(self, components=None):
+        # type: ([str, Component]) -> None
```
Comment (golanpa) [inline media_platform/service/flow_control_service/flow.py]:
```suggestion
        # type: (str, Component) -> None
```

File: media_platform/job/transcode/audio_qualities.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-class AudioQualities(object):
+class AudioQuality(object):
```
Comment (golanpa) [inline media_platform/job/transcode/audio_qualities.py]: consider using enums.

File: media_platform/service/flow_control_service/component.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from media_platform.job.extract_archive_job import ExtractArchiveSpecification
+from media_platform.job.extract_poster_job import ExtractPosterSpecification
+from media_platform.job.extract_storyboard_job import ExtractStoryboardSpecification
+from media_platform.job.import_file_job import ImportFileSpecification
+from media_platform.job.specification import Specification
+from media_platform.job.transcode_job import TranscodeSpecification
+from media_platform.lang.serialization import Deserializable, Serializable
+
+
+class ComponentType(object):
```
Comment (golanpa) [inline media_platform/service/flow_control_service/component.py]: consider using enum

File: media_platform/service/flow_control_service/component.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            'successors': self.successors,
+            'specification': self.specification.serialize() if self.specification else None,
+            'deleteSources': self.delete_sources
+        }
+
+    @classmethod
+    def deserialize(cls, data):
+        # type: (dict) -> Component
+        component_type = data['type']
+        specification_type = _SPECIFICATIONS[component_type]
```
Comment (golanpa) [inline media_platform/service/flow_control_service/component.py]: extract method: get_specification_for_component

File: media_platform/service/flow_control_service/component.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    ComponentType.transcode: TranscodeSpecification,
+    ComponentType.extract_poster: ExtractPosterSpecification,
+    ComponentType.extract_storyboard: ExtractStoryboardSpecification,
+    ComponentType.playlist: None,
+    ComponentType.import_file: ImportFileSpecification,
+}
+
+
+class Component(Serializable, Deserializable):
+    def __init__(self, component_type, successors, specification=None, delete_sources=False):
```
Comment (golanpa) [inline media_platform/service/flow_control_service/component.py]:
```suggestion
    def __init__(self, component_type, successors_ids, specification=None, delete_sources=False):
```

File: media_platform/service/flow_control_service/flow.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+        self.components = components or dict()
+
+    def add_component(self, key, component):
```
Comment (golanpa) [inline media_platform/service/flow_control_service/flow.py]: what is key? is it component key? i did not see that component has a key field? is it needed?

File: media_platform/service/flow_control_service/flow.py (hunk 1 above)
Comment (Elad-Laufer) [reply media_platform/service/flow_control_service/flow.py]: Single argument that is a dict, not 2 arguments...

File: media_platform/job/transcode/audio_qualities.py (hunk 2 above)
Comment (Elad-Laufer) [reply media_platform/job/transcode/audio_qualities.py]: I find proper enum restrictive and a hassle when dealing with forward compatibility

File: media_platform/service/flow_control_service/component.py (hunk 3 above)
Comment (Elad-Laufer) [reply media_platform/service/flow_control_service/component.py]: I find proper enum restrictive and a hassle when dealing with forward compatibility

File: media_platform/service/flow_control_service/component.py (hunk 4 above)
Comment (Elad-Laufer) [reply media_platform/service/flow_control_service/component.py]: What is its use? And how is it preferable to direct lookup?

File: media_platform/service/flow_control_service/component.py (hunk 5 above)
Comment (Elad-Laufer) [reply media_platform/service/flow_control_service/component.py]: successors do not have ids, those are keys.

File: media_platform/service/flow_control_service/flow.py (hunk 6 above)
Comment (Elad-Laufer) [reply media_platform/service/flow_control_service/flow.py]: Flow is a DAG, represented such:
{
 'a': [b,c],
 'b': [c],
 'c': []
}

In this case the successors (neighbours) are encapsulated in the component

File: media_platform/service/flow_control_service/flow.py (hunk 1 above)
Comment (alon-ne) [reply media_platform/service/flow_control_service/flow.py]: You can do it like this: dict[str, Component]

File: media_platform/service/flow_control_service/flow_invocation_request.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    def _params(self):
+        return {
+            'invocation': self.invocation.serialize(),
+            'flow': self.flow.serialize(),
+        }
+
+    def _validate_entry_points(self):
+        invalid_entry_points = [e for e in self.invocation.entry_points if e not in self.flow.components]
+        if invalid_entry_points:
+            raise ValueError('entry points not defined in components: ' % invalid_entry_points)
```
Comment (alon-ne) [inline media_platform/service/flow_control_service/flow_invocation_request.py]: Missing %s

**Pull request merged**